### ISSUE TITLE
shap() does not reconstruct predict(): values are negated/unscaled under SGD, and unrecoverable under Adam

### ISSUE THREAD

**Issue by ccaadaro** [NVlabs/gbrl#26]
## Summary

The ensemble-level `shap()` does not satisfy local accuracy (completeness), the defining
property of Shapley attributions:

```
predict(x) == base + sum_f shap(x)[f]
```

Two distinct symptoms, which I believe share one root cause:

1. **Under SGD**, completeness holds only after multiplying the returned values by `-lr`.
   The values come back **negated** and **unscaled by the learning rate**. Used as
   returned, every attribution has the **wrong sign** — a feature that suppressed a
   prediction appears to have driven it.

2. **Under Adam**, no constant rescale recovers completeness. The attributions do not
   reconstruct the model at all (best-fit residuals ~0.2–0.4 on a model whose predictions
   span ~±3).

Since SHAP is a property of a *trained* tree ensemble, the result should not depend on
which optimizer produced the leaf values. That it does is the bug.

Affects **1.1.7 and 1.1.8** (`src/cpp/shap.cpp` and `src/cpp/shap.h` are byte-identical
between the two releases).

## Reproducer

Self-contained, no external deps beyond `gbrl`/`torch`/`numpy`:

```python
import numpy as np, torch
from gbrl.models.gbt import GBTModel

def fit(algo, lr, output_dim=1, iterations=60, seed=0):
    rng = np.random.default_rng(seed)
    X = rng.normal(size=(400, 6)).astype(np.float32)
    Y = np.stack([2*X[:,0] - X[:,1],
                  X[:,2] + 0.5*X[:,3],
                  -X[:,0] + X[:,4]], axis=1)[:, :output_dim].astype(np.float32)

    model = GBTModel(
        input_dim=6, output_dim=output_dim,
        tree_struct={'max_depth': 4, 'n_bins': 256,
                     'min_data_in_leaf': 1, 'grow_policy': 'oblivious'},
        optimizers=[{'algo': algo, 'lr': lr, 'start_idx': 0, 'stop_idx': output_dim}],
        params={'split_score_func': 'Cosine', 'generator_type': 'Quantile'},
        device='cpu', verbose=0,
    )
    model.set_bias_from_targets(Y)

    target = torch.as_tensor(Y)
    for _ in range(iterations):
        pred = model(X, requires_grad=True)
        # reshape: GBRL squeezes the output to (n,) when output_dim == 1
        loss = ((pred.reshape(target.shape) - target) ** 2).mean()
        loss.backward()
        model.step(X)

    out = model(X).detach().cpu().numpy().reshape(len(X), output_dim)
    shap = np.asarray(model.shap(X)).reshape(len(X), 6, output_dim)
    return out, shap

def report(algo, lr, output_dim):
    pred, shap = fit(algo, lr, output_dim)
    print(f"\n{algo} lr={lr} output_dim={output_dim}")
    for k in range(output_dim):
        s, p = shap[:, :, k].sum(axis=1), pred[:, k]
        slope, intercept = np.polyfit(s, p, 1)
        # allow an arbitrary affine rescale per output -- the best case for completeness
        err = np.abs(slope * s + intercept - p).max()
        print(f"  output {k}: corr={np.corrcoef(s, p)[0,1]:+.6f}  "
              f"best-fit slope={slope:+.5f} (-lr={-lr:+.5f})  max residual={err:.2e}")

report("SGD",  0.1,  1)
report("SGD",  0.05, 1)
report("Adam", 0.01, 1)
report("Adam", 0.01, 3)
```

### Output (gbrl 1.1.7, CPU)

```
SGD lr=0.1 output_dim=1
  output 0: corr=-1.000000  best-fit slope=-0.10000 (-lr=-0.10000)  max residual=1.55e-04

SGD lr=0.05 output_dim=1
  output 0: corr=-1.000000  best-fit slope=-0.05000 (-lr=-0.05000)  max residual=1.30e-04

Adam lr=0.01 output_dim=1
  output 0: corr=-0.940253  best-fit slope=-0.00257 (-lr=-0.01000)  max residual=3.55e-01

Adam lr=0.01 output_dim=3
  output 0: corr=-0.924480  best-fit slope=-0.00773 (-lr=-0.01000)  max residual=4.28e-01
  output 1: corr=-0.990331  best-fit slope=-0.02093 (-lr=-0.01000)  max residual=1.93e-01
  output 2: corr=-0.956690  best-fit slope=-0.01207 (-lr=-0.01000)  max residual=2.75e-01
```

Expected in all cases: `corr = +1.000000`, `slope = +1.0`, residual ~0.

Observed:
- **SGD** — `corr` is exactly `-1`, and the best-fit slope is exactly `-lr` at every
  learning rate tried. So the values are the true contributions scaled by `-1/lr`.
  Recoverable, but only if you know to apply `-lr`, which is not documented.
- **Adam** — the slope no longer matches `-lr` and varies per output; residuals are
  ~10–20% of the prediction range. Not recoverable by any constant factor.

## Suspected root cause

`shap()` looks like it decomposes the **pre-optimizer leaf targets** (the fitted
gradients) rather than the leaf values actually written into the ensemble.

That is consistent with both symptoms:

- Plain SGD applies a constant step `-lr * grad`, so the decomposition is off by exactly
  a factor of `-lr` — which is what the reproducer measures, to 5 decimal places.
- Adam applies a per-coordinate adaptive step, so there is no single factor relating the
  fitted gradients to the stored leaf values, and completeness cannot be restored by any
  rescale.

If that is right, the fix is for `shap()` to decompose the ensemble's actual leaf values
(post-optimizer), which would make it optimizer-independent and remove the need for any
caller-side correction.

## Why this went unnoticed

`tests/test_gbt_single.py::test_shap_cpu` is the only SHAP test, and it exercises a
different code path:

```python
model.learner.step(X, y)                              # a single boosting step -> 1 tree
gbrl_shap = model.tree_shap(0, X_cpu[0, :])[0].flatten()   # per-tree, not the ensemble
clf = DecisionTreeRegressor(max_depth=3).fit(X_cpu, y)
target_shap = shap.TreeExplainer(clf).shap_values(X_cpu[0])
```

It covers `tree_shap` on a **single tree**, with **SGD**, after **one** boosting
iteration, with **one** output. The ensemble-level `shap()` is never tested — not for
completeness, not across multiple trees, not with a non-SGD optimizer, and not with
`output_dim > 1`. A completeness assertion (`predict == base + sum(shap)`) over a
multi-tree ensemble would catch both symptoms above.

## Impact

Interpretability is a headline use case for gradient-boosted RL policies, and this is the
API people reach for. Used as documented:

- On the **SGD** path — which is `gbrl_sb3`'s default (`algos/ppo.py`: `'policy_algo': 'SGD'`)
  — every attribution comes out **sign-inverted**, so conclusions are backwards.
- On the **Adam** path — an optimizer the API offers and which users do select — the
  attributions do not correspond to the model at all, and no rescale makes them.

Neither failure is announced: `shap()` returns plausible-looking numbers in both cases.

We hit this building a SHAP-based analysis of a GBRL PPO policy (trained via `gbrl_sb3`
with Adam). The attributions failed a completeness check, which is the only reason we
caught it; without that check we would have published an explanation that does not
correspond to the trained model.

## Suggested fixes

1. Have `shap()` decompose the ensemble's stored leaf values so it is optimizer-independent
   and satisfies `predict == base + sum(shap)` as returned.
2. Failing that (or in the interim): document the `-lr` sign/scale convention, and raise
   or warn when `shap()` is called on a model trained with a non-constant-step optimizer,
   rather than returning values that silently do not reconstruct the model.
3. Add a completeness regression test over a multi-tree, multi-output ensemble for each
   supported optimizer.

## Environment

- gbrl 1.1.7 (verified unchanged in 1.1.8), CPU
- Python 3.11, numpy 2.x, torch 2.x
- Linux

**Comment by benja263** [comment]
Thank you for pointing this out. This issue has now been resolved in GBRL 1.1.9.
